src/Popover.js

```diff
@@ -1,79 +1,166 @@
+    PropTypes.string,
+    PropTypes.func,
+    DOMElement,
+  ]).isRequired,
   isOpen: PropTypes.bool,
-  tether: PropTypes.object,
-  tetherRef: PropTypes.func,
+  disabled: PropTypes.bool,
   className: PropTypes.string,
+  placementPrefix: PropTypes.string,
   cssModule: PropTypes.object,
   toggle: PropTypes.func,
+  delay: PropTypes.oneOfType([
+    PropTypes.shape({ show: PropTypes.number, hide: PropTypes.number }),
+    PropTypes.number,
+  ]),
 };
 
-const defaultProps = {
-  isOpen: false,
-  placement: 'bottom',
-  toggle: () => {}
+const DEFAULT_DELAYS = {
+  show: 0,
+  hide: 0,
 };
 
-const defaultTetherConfig = {
-  classPrefix: 'bs-tether',
-  classes: {
-    element: false,
-    enabled: 'show',
-  },
-  constraints: [
-    { to: 'scrollParent', attachment: 'together none' },
-    { to: 'window', attachment: 'together none' }
-  ]
+const defaultProps = {
+  isOpen: false,
+  placement: 'right',
+  placementPrefix: 'bs-popover',
+  delay: DEFAULT_DELAYS,
+  toggle: () => {},
 };
 
 class Popover extends React.Component {
   constructor(props) {
     super(props);
 
-    this.getTetherConfig = this.getTetherConfig.bind(this);
+    this.addTargetEvents = this.addTargetEvents.bind(this);
+    this.handleDocumentClick = this.handleDocumentClick.bind(this);
+    this.removeTargetEvents = this.removeTargetEvents.bind(this);
+    this.toggle = this.toggle.bind(this);
+    this.show = this.show.bind(this);
+    this.hide = this.hide.bind(this);
+  }
+
+  componentDidMount() {
+    this._target = getTarget(this.props.target);
+    this.handleProps();
   }
 
-  getTetherConfig() {
-    const attachments = getTetherAttachments(this.props.placement);
-    return {
-      ...defaultTetherConfig,
-      ...attachments,
-      target: '#' + this.props.target,
-      ...this.props.tether
-    };
+  componentDidUpdate() {
+    this.handleProps();
+  }
+
+  componentWillUnmount() {
+    this.clearShowTimeout();
+    this.clearHideTimeout();
+    this.removeTargetEvents();
+  }
+
+  getDelay(key) {
+    const { delay } = this.props;
+    if (typeof delay === 'object') {
+      return isNaN(delay[key]) ? DEFAULT_DELAYS[key] : delay[key];
+    }
+    return delay;
+  }
+
+  handleProps() {
+    if (this.props.isOpen) {
+      this.show();
+    } else {
+      this.hide();
+    }
+  }
+
+  show() {
+    this.clearHideTimeout();
+    this.addTargetEvents();
+    if (!this.props.isOpen) {
+      this.clearShowTimeout();
+      this._showTimeout = setTimeout(this.toggle, this.getDelay('show'));
+    }
+  }
+
+  hide() {
+    this.clearShowTimeout();
+    this.removeTargetEvents();
+    if (this.props.isOpen) {
+      this.clearHideTimeout();
+      this._hideTimeout = setTimeout(this.toggle, this.getDelay('hide'));
+    }
+  }
+
+  clearShowTimeout() {
+    clearTimeout(this._showTimeout);
+    this._showTimeout = undefined;
+  }
+
+  clearHideTimeout() {
+    clearTimeout(this._hideTimeout);
+    this._hideTimeout = undefined;
+  }
+
```

Comment:
Will all the show/hide timeout logic ... will this also fix the few issues we have with fast enable/disabling of popovers/tooltips where it was previously crashing? If not, maybe test with delays of 2000 (2s) to make sure we can close those issues too.

Reply:
Animations still need to be reworked, but the way this works it should better handle timeouts and toggles. I don't really want to have a long running test; there are tests to ensure the timeouts are cleared.